### PR TITLE
Add TokenPool contract and it's tests

### PR DESCRIPTION
**Pull request by harzo**
Unused imports in tokenpool.test.ts left on purpose.

### CODE REVIEW

**Comment by harzo** [comment]
@jstefanski Please, review changes.

File: test/tokenpool.test.ts [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    it('should revert for zero amount', async () => {
+      await assertReverts(async () => {
+        await tokenPool.registerPool(poolName, 0, {
+          from: owner
+        });
+      });
+    });
+
+    it('should revert for not unique pool', async () => {
+      await tokenPool.registerPool(poolName, poolTokenAmount, {
```
Comment (harzo) [inline test/tokenpool.test.ts]: In Here, I could create new context but IMHO its to much for only one test.
@jstefanski what do you think?

Comment (jstefanski) [reply test/tokenpool.test.ts]: Yes, separate contexts are only justified if there are several test cases that need some initial setup.

**Comment by jstefanski** [comment]
I fixed solhint and prettier scripts. Previously it interpreted wildcards incorrectly and skipped some files. Run `npm run format` and `npm run lint` now and fix errors in code.

**Comment by harzo** [comment]
@jstefanski All linter errors fixed. Please confirm that all is good. Thanks

File: contracts/TokenPool.sol [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    modifier onlyUnlocked(string pool) {
+        /* solhint-disable not-rely-on-time */
+        require(now > poolRegister[pool].lockTimestamp);
+        /* solhint-enable not-rely-on-time */
+        _;
+    }
+
+    modifier onlyUnique(string pool) {
```
Comment (jstefanski) [inline contracts/TokenPool.sol]: Rename to `unlyUniquePool()` - passed argument (string) is not directly checked. It looks similar to `onlyValid()` and `onlyNotZero()` which just check the passed argument without accessing storage.

File: contracts/TokenPool.sol [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        _;
+    }
+
+    modifier onlyValid(address _address) {
+        require(_address != 0);
+        _;
+    }
+
+    function TokenPool(Mintable _token)
+        public
```
Comment (jstefanski) [inline contracts/TokenPool.sol]: Put `public` modifier in the same line. It is far from exceeding 120 columns as set in `.solhint`.

File: contracts/TokenPool.sol [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+     */
+    function transfer(address to, string pool, uint256 amount)
+        public
+        onlyOwner
+        onlyValid(to)
+        onlyNotZero(amount)
+        onlySufficientAmount(pool, amount)
+        onlyUnlocked(pool)
+    {
+        token.transfer(to, amount);
```
Comment (jstefanski) [inline contracts/TokenPool.sol]: Check `ERC20.transfer()` result

File: contracts/TokenPool.sol [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        poolRegister[pool].amount -= amount;
```
Comment (jstefanski) [inline contracts/TokenPool.sol]: Use `SafeMath` `sub`. I know this is protected by precondition `onlySufficientAmount` but additional assertion won't hurt.

File: contracts/TokenPool.sol [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        public
+        onlyOwner
+        onlyNotZero(amount)
+        onlyUnique(name)
+    {
+        poolRegister[name] = Pool(amount, 0);
+        token.mint(this, amount);
+        PoolRegistered(name, amount);
+
+        if (lockTimestamp > 0) lockPool(name, lockTimestamp);
```
Comment (jstefanski) [inline contracts/TokenPool.sol]: Don't use one-line `if` statements, it's prone to errors

**Pull request merged**